### PR TITLE
Add support for torch 2.0

### PR DESCRIPTION
**Pull request by dakinggg**
# What does this PR do?
This PR upgrades the torch pin to support torch 2.0. It includes related fixes, mostly resulting from the use of `use_orig_params=True` with FSDP, which is necessary to support `compile`.

Changes:
- Different way of loading FSDP state dicts for torch 2
- `_LRScheduler` -> `LRScheduler`
- using `summon_full_params` to get HF generate to work with FSDP
- fixing the way we compute optimizer metrics to allow for the fact that not all ranks have all params
- assorted minor fixes and test fixes

Manual tests:
- [x] sharded multinode autoresume
Screen Shot 2023-04-26 at 7 37 10 PM

- [x] local autoresume
Screen Shot 2023-04-26 at 7 36 33 PM

- [x] full resume
Screen Shot 2023-04-26 at 7 35 59 PM

- [x] optimizer monitor (https://wandb.ai/mosaic-ml/daniel-debug/runs/34o29laf?workspace=user-danielking)
- [x] base 1b run (https://wandb.ai/mosaic-ml/daniel-debug/runs/37e2g60h?workspace=user-danielking)
- [x] hf generate for gpt neo (https://wandb.ai/mosaic-ml/daniel-debug/groups/torch2-evals/workspace?workspace=user-danielking)
- [x] hf generate for mosaicgpt (https://wandb.ai/mosaic-ml/daniel-debug/runs/10lknf0v?workspace=user-danielking)
- [x] added remote tests pass
Screen Shot 2023-04-27 at 12 49 25 AM

# What issue(s) does this change relate to?
Closes [CO-2029](https://mosaicml.atlassian.net/browse/CO-2029)
Closes https://github.com/mosaicml/composer/issues/2147

# Before submitting
- [x] Have you read the [contributor guidelines](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md)?
- [x] Was this change discussed/approved in a GitHub issue first? It is much more likely to be merged if so.
- [x] Did you update any related docs and document your change?
- [x] Did you update any related tests and add any new tests related to your change? (see [testing](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#running-tests))
- [x] Did you run the tests locally to make sure they pass?
- [x] Did you run `pre-commit` on your change? (see the `pre-commit` section of [prerequisites](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#prerequisites))

[CO-2029]: https://mosaicml.atlassian.net/browse/CO-2029?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: .github/workflows/pr-gpu.yaml [hunk 1]
```diff
@@ -1,6 +1,6 @@
 name: PR GPU tests
 on:
-  pull_request_target:
+  pull_request:
```
Comment (dakinggg) [inline .github/workflows/pr-gpu.yaml:3]: don't commit this

File: .github/workflows/pr-cpu.yaml [hunk 2]
```diff
@@ -36,6 +36,11 @@ jobs:
             markers: 'not daily and not remote and not gpu and not vision and not doctest'
             pytest_command: 'coverage run -m pytest'
             composer_package_name: 'composer'
+          - name: 'cpu-3.10-2.0-composer'
```
Comment (mvpatel2000) [inline .github/workflows/pr-cpu.yaml:39]: can you update daily tests to include this please?

File: .github/workflows/pr-gpu.yaml [hunk 3]
```diff
@@ -21,6 +21,11 @@ jobs:
             markers: 'not daily and not remote and gpu and (vision or not vision) and (doctest or not doctest)'
             pytest_command: 'coverage run -m pytest'
             composer_package_name: 'mosaicml'
+          - name: 'gpu-3.10-2.0-vision-doctest'
```
Comment (mvpatel2000) [inline .github/workflows/pr-gpu.yaml:24]: can you update daily tests to include this please?

File: composer/callbacks/optimizer_monitor.py [hunk 4]
```diff
@@ -99,7 +99,7 @@ def batch_end(self, state: State, logger: Logger):
 
                 metric_reporter = getattr(state.optimizers[0], 'report_per_parameter_metrics', None)
                 if callable(metric_reporter) and self.log_optimizer_metrics:
-                    optimizer_metrics = metric_reporter(p, name, optimizer_metrics)
+                    optimizer_metrics.update(metric_reporter(p, name, optimizer_metrics))
```
Comment (mvpatel2000) [inline composer/callbacks/optimizer_monitor.py:102]: do we need to pass in `optimizer_metrics` as an arg if we're calling `update`?

File: composer/core/state.py [hunk 5]
```diff
@@ -27,6 +27,7 @@
 from composer.core.time import Time, Timestamp, TimeUnit
 from composer.devices import Device
 from composer.utils import batch_get, batch_set, dist, ensure_tuple, get_composer_env_dict, is_model_deepspeed
+from composer.utils.misc import using_torch_2_0
```
Comment (mvpatel2000) [inline composer/core/state.py:30]: can we rename this to `using_torch_2` since presumably we gate for >= 2.0?

File: composer/core/state.py [hunk 6]
```diff
@@ -132,6 +147,32 @@ def fsdp_get_optim_state_dict(model: torch.nn.Module,
         raise NotImplementedError(f'No valid FSDP state_dict_type for {state_dict_type}')
 
 
+def _legacy_optim_state_dict_to_load(optim_state_dict: Dict[str, Any],
+                                     model: torch.nn.Module,
+                                     optim: torch.optim.Optimizer,
+                                     state_dict_type: str = 'full'):
```
Comment (mvpatel2000) [inline composer/core/state.py:153]:
```suggestion
                                     state_dict_type: str = 'full',):
```
nit. I'll add a lint for this soon...

File: composer/core/types.py [hunk 7]
```diff
@@ -27,7 +27,10 @@
 
 Dataset = torch.utils.data.Dataset[Batch]
 
-PyTorchScheduler = torch.optim.lr_scheduler._LRScheduler
+try:
+    PyTorchScheduler = torch.optim.lr_scheduler.LRScheduler
+except:
+    PyTorchScheduler = torch.optim.lr_scheduler._LRScheduler
```
Comment (mvpatel2000) [inline composer/core/types.py:33]: can we leave a comment on what versions require what? eventually (likely a long time) we want to deprecate, and it's useful to know when

File: composer/models/huggingface.py [hunk 8]
```diff
@@ -465,8 +470,17 @@ def generate(self, input_ids: torch.Tensor, **kwargs):
...
             self.dummy_forward_called = True
 
-        pad_token_id = kwargs.pop('pad_token_id', self.tokenizer.pad_token_id if self.tokenizer is not None else None)
-        return self.model.generate(input_ids, pad_token_id=pad_token_id, **kwargs)
+        if is_model_fsdp(self.model) and using_torch_2_0():
+            from torch.distributed.fsdp import FullyShardedDataParallel as FSDP
+
+            # NOTE: We need to use the FSDP.summon_full_params context manager here because the generate function
+            # does not seem to gather the weights for the LM head. This has only been tested with a few models, so
+            # consider FSDP + Torch 2.0 + HuggingFace generate to be beta suppor.
```
Comment (mvpatel2000) [inline composer/models/huggingface.py:478]:
```suggestion
            # consider FSDP + Torch 2.0 + HuggingFace generate to be beta support.
```

File: composer/models/huggingface.py [hunk 9]
```diff
@@ -451,10 +451,15 @@ def generate(self, input_ids: torch.Tensor, **kwargs):
...
         """
+        pad_token_id = kwargs.pop('pad_token_id', self.tokenizer.pad_token_id if self.tokenizer is not None else None)
+
+        from composer.utils.misc import using_torch_2_0
+
         # We need to call forward once in order for FSDP + generate to work
         # See https://github.com/huggingface/accelerate/issues/570, https://github.com/huggingface/accelerate/issues/947,
         # and https://github.com/pytorch/pytorch/issues/82461 for more info
-        if not self.dummy_forward_called and is_model_fsdp(self.model):
+        # NOTE: This is a hack for Torch 1.13.x, and there is a different hack below for Torch 2.0
```
Comment (mvpatel2000) [inline composer/models/huggingface.py:461]:
```suggestion
        # Note: This is a hack for Torch 1.13.x, and there is a different hack below for Torch 2.0
```
less aggro :P

File: composer/models/huggingface.py [hunk 10]
```diff
@@ -465,8 +470,17 @@ def generate(self, input_ids: torch.Tensor, **kwargs):
                            **maybe_decoder_input_ids)
             self.dummy_forward_called = True
 
-        pad_token_id = kwargs.pop('pad_token_id', self.tokenizer.pad_token_id if self.tokenizer is not None else None)
-        return self.model.generate(input_ids, pad_token_id=pad_token_id, **kwargs)
+        if is_model_fsdp(self.model) and using_torch_2_0():
+            from torch.distributed.fsdp import FullyShardedDataParallel as FSDP
+
+            # NOTE: We need to use the FSDP.summon_full_params context manager here because the generate function
```
Comment (mvpatel2000) [inline composer/models/huggingface.py:476]:
```suggestion
            # Note: We need to use the FSDP.summon_full_params context manager here because the generate function
```

File: composer/models/huggingface.py [hunk 11]
```diff
@@ -465,8 +470,17 @@ def generate(self, input_ids: torch.Tensor, **kwargs):
...
+            # does not seem to gather the weights for the LM head. This has only been tested with a few models, so
+            # consider FSDP + Torch 2.0 + HuggingFace generate to be beta suppor.
+            # NOTE: We use recurse=False here so that we only summon full params for the LM head, not the entire model.
```
Comment (mvpatel2000) [inline composer/models/huggingface.py:479]:
```suggestion
            # Note: We use recurse=False here so that we only summon full params for the LM head, not the entire model.
```

File: composer/utils/auto_log_hparams.py [hunk 12]
```diff
@@ -20,8 +20,9 @@ def extract_hparams(locals_dict: Dict[str, Any]) -> Dict[str, Any]:
             value or to another sub_dict.
     """
     hparams = {}
+    print(locals_dict)
```
Comment (mvpatel2000) [inline composer/utils/auto_log_hparams.py:23]: remove?

File: composer/utils/module_surgery.py [hunk 13]
```diff
@@ -371,7 +371,9 @@ def update_params_in_optimizer(old_params: Iterable[torch.nn.parameter.Parameter
 
     # rip out the removed_params' states from the optimizer
     for p in removed_params:
-        if _tensor_in(p, opt.state):  # only true after training starts
+        # only true after training starts
+        # Note, it appears that opt.state is a dict whose keys _might_ be tensors
```
Comment (mvpatel2000) [inline composer/utils/module_surgery.py:375]:
```suggestion
        # Only true after training starts
        # Note: it appears that opt.state is a dict whose keys _might_ be tensors
```

File: tests/algorithms/test_gradient_clipping.py [hunk 14]
```diff
@@ -192,12 +193,27 @@ def test_gradient_clipping_algorithm_with_deepspeed_enabled(
...
+        return False
+else:
+
+    def _auto_wrap_policy(module: torch.nn.Module, recurse: bool, nonwrapped_numel: int) -> bool:
+        if recurse:
+            return True
+
+        # With Torch 2.0, there is a bug that emits a nasty warning if you wrap a module with no parameters
+        if len(list(module.parameters())) == 0:
+            return False
```
Comment (mvpatel2000) [inline tests/algorithms/test_gradient_clipping.py:212]: Do we need to special case this w 2.0 or can we always do this?

File: tests/callbacks/test_optimizer_monitor.py [hunk 15]
```diff
@@ -71,29 +76,116 @@ def test_fsdp_optimizer_monitor(device, world_size):
                       optimizers=DecoupledAdamW(model.parameters()),
                       max_duration='3ba',
                       fsdp_config={
-                          'sharding_strategy': 'FULL_SHARD',
-                          'min_params': 10,
+                          'sharding_strategy': 'FULL_SHARD' if world_size > 1 else 'NO_SHARD',
+                          'min_params': 1,
```
Comment (mvpatel2000) [inline tests/callbacks/test_optimizer_monitor.py:80]: this is about to get removed iirc / is deprecated. Can be separate PR to fix? CC: @bcui19

File: tests/common/models.py [hunk 16]
```diff
@@ -23,13 +23,18 @@ class SimpleModel(ComposerClassifier):
         num_classes (int): number of classes (default: 2)
     """
 
-    def __init__(self, num_features: int = 1, num_classes: int = 2, device: str = 'cpu', bias: bool = True) -> None:
+    def __init__(self,
+                 num_features: int = 1,
+                 num_classes: int = 2,
+                 num_hidden: int = 8,
+                 device: str = 'cpu',
+                 bias: bool = True) -> None:
```
Comment (mvpatel2000) [inline tests/common/models.py:31]:
```suggestion
                 bias: bool = True,) -> None:
```

File: tests/trainer/test_sharded_checkpoint.py [hunk 17]
```diff
@@ -5,19 +5,33 @@
...
+from composer.models import ComposerClassifier
 from composer.trainer import Trainer
 from composer.utils import dist
-from tests.common import RandomClassificationDataset, SimpleModel
+from composer.utils.file_helpers import get_file
+from tests.common import RandomClassificationDataset
 from tests.common.markers import world_size
 
 
+# This model is to be used explicitly for this unit test because it some old reference checkpoints
```
Comment (mvpatel2000) [inline tests/trainer/test_sharded_checkpoint.py:22]:
```suggestion
# This model is to be used explicitly for this unit test because some old reference checkpoints
```

File: tests/trainer/test_sharded_checkpoint.py [hunk 18]
```diff
@@ -240,6 +165,48 @@ def test_fsdp_full_state_dict_load(world_size, tmp_path: pathlib.Path, autoresum
     trainer2.fit()
 
 
+@pytest.mark.gpu
+@pytest.mark.remote
```
Comment (mvpatel2000) [inline tests/trainer/test_sharded_checkpoint.py:169]: can u post a screenshot showing this passes as it only runs on daily?

**Review by mvpatel2000 — COMMENTED** [review]
Why is there a gap in the resumption tests?

Mostly LGTM / minor comments. will do one more pass after comments are resolved before approval because PR is massive

File: composer/callbacks/optimizer_monitor.py (hunk 4 above)
Comment (dakinggg) [reply composer/callbacks/optimizer_monitor.py:102]: yes, the `metric_reporter` function has loops over `optimizer_metrics` and such.

File: tests/callbacks/test_optimizer_monitor.py (hunk 15 above)
Comment (dakinggg) [reply tests/callbacks/test_optimizer_monitor.py:80]: I think I can just remove it tbh, might need to set some `_fsdp_wrap` or something.

File: tests/algorithms/test_gradient_clipping.py (hunk 14 above)
Comment (dakinggg) [reply tests/algorithms/test_gradient_clipping.py:212]: Unfortunately there is also a parameter renamed between these two functions.

File: tests/trainer/test_sharded_checkpoint.py (hunk 18 above)
Comment (dakinggg) [reply tests/trainer/test_sharded_checkpoint.py:169]: lol lets hope @eracah ran it, but yeah, will run and include

File: composer/utils/auto_log_hparams.py (hunk 12 above)
Comment (dakinggg) [reply composer/utils/auto_log_hparams.py:23]: oops, will do

File: composer/models/huggingface.py [hunk 19]
```diff
@@ -465,8 +470,17 @@ def generate(self, input_ids: torch.Tensor, **kwargs):
...
-        pad_token_id = kwargs.pop('pad_token_id', self.tokenizer.pad_token_id if self.tokenizer is not None else None)
-        return self.model.generate(input_ids, pad_token_id=pad_token_id, **kwargs)
+        if is_model_fsdp(self.model) and using_torch_2_0():
+            from torch.distributed.fsdp import FullyShardedDataParallel as FSDP
+
+            # Note: We need to use the FSDP.summon_full_params context manager here because the generate function
+            # does not seem to gather the weights for the LM head. This has only been tested with a few models, so
+            # consider FSDP + Torch 2.0 + HuggingFace generate to be beta support.
+            # Note: We use recurse=False here so that we only summon full params for the LM head, not the entire model.
+            with FSDP.summon_full_params(self.model, writeback=False, recurse=False):
```
Comment (dakinggg) [inline composer/models/huggingface.py:480]: adjust comment after discussing with pytorch and link to relevant issue

File: .github/workflows/pr-cpu.yaml [hunk 20]
```diff
@@ -36,6 +36,11 @@ jobs:
...
+            container: mosaicml/pytorch:2.0.0_cpu-python3.10-ubuntu20.04
+            markers: 'not daily and not remote and not gpu and not vision and not doctest'
+            pytest_command: 'coverage run -m pytest'
+            composer_package_name: 'composer'
```
[the remark this answers is not in the document]
Comment (dakinggg) [reply .github/workflows/pr-cpu.yaml:43]: oh sure, i just copy pasted blindly from the top of the list

**Comment by dakinggg** [comment]
@mvpatel2000 the presumptions with a gap are because run 1 went for 10 batches, and then run 2 was run with autoresume true and an increased max duration, rather than deleting some checkpoints or something. Lmk if that makes sense.

File: .github/workflows/pr-cpu.yaml (hunk 2 above)
Comment (dakinggg) [reply .github/workflows/pr-cpu.yaml]: added

File: composer/core/state.py (hunk 5 above)
Comment (dakinggg) [reply composer/core/state.py]: done

File: .github/workflows/pr-gpu.yaml (hunk 3 above)
Comment (dakinggg) [reply .github/workflows/pr-gpu.yaml:24]: it doesn't look like we have any gpu tests in the daily? am i missing something?

File: tests/trainer/test_sharded_checkpoint.py (hunk 18 above)
Comment (dakinggg) [reply tests/trainer/test_sharded_checkpoint.py:182]: done

File: composer/core/types.py (hunk 7 above)
Comment (dakinggg) [reply composer/core/types.py]: done

File: composer/utils/auto_log_hparams.py (hunk 12 above)
Comment (dakinggg) [reply composer/utils/auto_log_hparams.py]: done

File: composer/core/state.py [hunk 21]
```diff
@@ -956,29 +1006,22 @@ def load_optim_state(self, state_dict: Dict[str, Any]):
                 continue
             optim_state_dict = serialized_value[type(optimizer).__qualname__]
             if self.fsdp_enabled:
+                assert self.fsdp_state_dict_type is not None
```
Comment (karan6181) [inline composer/core/state.py:1009]: nit: raise `ValueError` or any other Exception type with a descriptive error message instead of `assert` ?

File: composer/core/types.py [hunk 22]
```diff
@@ -27,7 +27,12 @@
 
 Dataset = torch.utils.data.Dataset[Batch]
 
-PyTorchScheduler = torch.optim.lr_scheduler._LRScheduler
+try:
```
Comment (karan6181) [inline composer/core/types.py:30]: Why not having a `if using_torch_2()` check instead ?

File: composer/utils/auto_log_hparams.py [hunk 23]
```diff
@@ -21,7 +21,7 @@ def extract_hparams(locals_dict: Dict[str, Any]) -> Dict[str, Any]:
     """
     hparams = {}
     for k, v in locals_dict.items():
-        if k.startswith('_') or k == 'self':
+        if k.startswith('_') or k == 'self' or type(v) is type:
```
Comment (karan6181) [inline composer/utils/auto_log_hparams.py:24]: what is `type(v) is type` lol ?

**Review by karan6181 — COMMENTED** [review]
Some minor comments. Overall looks good. I liked the detailed comments that you have added at multiple places to get better understanding of a code. 

Minor nit: More descriptive PR header name ?

File: composer/core/state.py (hunk 21 above)
Comment (dakinggg) [reply composer/core/state.py:1009]: this is just for pyright, will add a comment

File: .github/workflows/pr-gpu.yaml (hunk 3 above)
Comment (dakinggg) [reply .github/workflows/pr-gpu.yaml:24]: ah, thank you. done

**Review by karan6181 — APPROVED** [review]
LGTM. Thanks!

**Comment by dakinggg** [comment]
I think the message you get from ONNX directly is about as clear as it gets...since we don't know which operator they're having trouble with and how that correspond to opset version and torch version. but open to another suggestion
